### PR TITLE
2022.9.x feature privacy console

### PR DESCRIPTION
**Pull request by vicpatel**
Add data privacy request route and implementation

- Introduces new role for data-privacy-officer(Role allows user to manage data privacy requests)
- along with a new system resource for data privacy requests and its access control
- Routes as per access control to create data privacy request, list request and filter it based on their kind and status, and update data privacy request status

Please take a look.

### CODE REVIEW

**Comment by darh** [comment]
will those commit messages get some love? :)

File: pkg/rbac/roles.go [hunk 1]
```diff
@@ -55,7 +55,7 @@ func (k roleKind) String() string {
 	case AnonymousRole:
 		return "anonymous"
 	default:
-		panic("unknown role kind")
```
Comment (darh) [inline pkg/rbac/roles.go]: what, why?

File: pkg/wfexec/session.go [hunk 2]
```diff
@@ -612,7 +612,8 @@ func (s *Session) exec(ctx context.Context, log *zap.Logger, st *State) (nxt []*
 	st.created = *now()
 
 	defer func() {
-		reason := recover()
+		var reason interface{}
```
Comment (darh) [inline pkg/wfexec/session.go]: what does this have to do with what this PR is about?

File: store/adapters/rdbms/upgrade_tables.go [hunk 3]
```diff
@@ -696,3 +697,22 @@ func tableResourceActivityLog() *Table {
...
+		ID,
+		ColumnDef("name", ColumnTypeIdentifier),
+		ColumnDef("request_type", ColumnTypeText, ColumnTypeLength(handleLength)),
+		ColumnDef("status", ColumnTypeInteger),
+		ColumnDef("requested_at", ColumnTypeTimestamp),
+		ColumnDef("requested_by", ColumnTypeTimestamp),
+		ColumnDef("completed_at", ColumnTypeTimestamp),
+		ColumnDef("completed_by", ColumnTypeTimestamp),
+
+		AddIndex("request_type", IColumn("request_type")),
```
Comment (darh) [inline store/adapters/rdbms/upgrade_tables.go]: index overload? do we really need all of them?

**Review by darh — CHANGES_REQUESTED** [review]

File: store/adapters/rdbms/upgrade_tables.go (hunk 3 above)
Comment (vicpatel) [reply store/adapters/rdbms/upgrade_tables.go]: No

File: pkg/rbac/roles.go (hunk 1 above)
Comment (vicpatel) [reply pkg/rbac/roles.go]: Just IDE, fixing problems for me :D

File: pkg/wfexec/session.go (hunk 2 above)
Comment (vicpatel) [reply pkg/wfexec/session.go]: MB, It is not relevant to this PR.

File: compose/service/locale.go [hunk 4]
```diff
@@ -272,6 +272,43 @@ func (svc resourceTranslationsManager) pageExtended(ctx context.Context, res *ty
...
+				Lang:     tag.String(),
+				Key:      yAxisLabelK.Path,
+				Msg:      svc.locale.TResourceFor(tag, res.ResourceTranslation(), yAxisLabelK.Path),
+			})
+		}
+
+		for _, metric := range report.Metrics {
+			if _, ok := metric["metricID"]; ok {
+				mpl := strings.NewReplacer(
+					"{{metricID}}", metric["metricID"].(string),
```
Comment (tjerman) [inline compose/service/locale.go]: This feels a bit dangerous to me; if it's not ok it will panic
I see you do some validation before this, but I would still prefer you do the *more safe* type assertion.
Alternatively, you can add a bunch of tests to assure this will be fine.

Comment (tjerman) [reply compose/service/locale.go]: Apply same comment for the rest of such cases

File: pkg/provision/migrations_202203_resource_translations.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+// Migrates resource translations from the resource
+// struct to the dedicated store (table)
+//
+// While doing this, we also modify some resource substructure:
+//  - chart reports (assign report IDs)
+//
+// Note: we will migrate all translations to current default language
+// If you do not like that, shut down Corteza after migrations and fix this directly in the store
+func migratePost202203ResourceTranslations(ctx context.Context, log *zap.Logger, s store.Storer) (err error) {
+	log.Info("migrating post 202203 resource locales")
```
Comment (tjerman) [inline pkg/provision/migrations_202203_resource_translations.go]: `migrating post 202203 resource translations`

File: pkg/provision/migrations_202203_resource_translations.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+	var (
+		migrated = make(map[string]bool)
+	)
+
+	set, _, err := store.SearchResourceTranslations(ctx, s, sysTypes.ResourceTranslationFilter{})
```
Comment (tjerman) [inline pkg/provision/migrations_202203_resource_translations.go:30]: You missed the error check

File: pkg/wfexec/session.go [hunk 7]
```diff
@@ -612,7 +612,8 @@ func (s *Session) exec(ctx context.Context, log *zap.Logger, st *State) (nxt []*
...
+		reason = recover()
 		if reason == nil {
```
Comment (tjerman) [inline pkg/wfexec/session.go]: Why split it up?

File: system/data_privacy_request.cue [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+	struct: {
+		id: schema.IdField
+		kind: { goType: "types.RequestKind", sortable: true }
+		status: { goType: "types.RequestStatus", sortable: true }
+
+		requested_at: { schema.SortableTimestampField, ident: "requestedAt", storeIdent: "requested_at" }
+		requested_by: { goType: "uint64", ident: "requestedBy", storeIdent: "requested_by" }
+		completed_at: { schema.SortableTimestampNilField, ident: "completedAt", storeIdent: "completed_at" }
+		completed_by: { goType: "uint64", ident: "completedBy", storeIdent: "completed_by" }
```
Comment (tjerman) [inline system/data_privacy_request.cue]: I think you can omit all of the `ident` and `storeIdent` here

File: system/rest.yaml [hunk 9]
```diff
@@ -2198,3 +2198,56 @@ endpoints:
...
+        post:
+          - { name: kind,         type: "string",   title: "Request Kind", required: true }
+    - name: update status
+      method: PUT
+      title: Update data privacy request status
+      path: "/{requestID}/status/{status}"
+      parameters:
+        path:
+          - { name: requestID,     type: "uint64",   title: "ID",             required: true }
+          - { name: status,        type: "string",   title: "Request Status", required: true }
```
Comment (tjerman) [inline system/rest.yaml]: I think that `PATCH` would be more appropriate here since we're updating just the status

File: system/rest.yaml [hunk 10]
```diff
@@ -2198,3 +2198,56 @@ endpoints:
...
+      parameters:
+        path:
+          - { name: requestID, type: "string",     title: "Request ID", required: true }
+    - name: create response
+      method: GET
+      title: Create response for a specific request
+      path: "/{requestID}/responses"
+      parameters:
+        path:
+          - { name: requestID, type: "string",     title: "Request ID", required: true }
```
Comment (tjerman) [inline system/rest.yaml]: `POST` with a payload?

File: system/types/data_privacy.go [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+	RequestStatusRejected RequestStatus = "rejected"
+)
+
+func (k RequestKind) IsValid() error {
+	switch k {
+	case RequestKindCorrect, RequestKindDelete, RequestKindExport:
+		return nil
+	}
+	return errors.New("invalid request kind")
+}
```
Comment (tjerman) [inline system/types/data_privacy.go]: If you only want to allow these request kinds, what I would consider here is not exposing the `RequestKind` type and defining some conversion function (either here, in the `types` or where you're constructing the privacy struct).

If you hide the type you don't need to worry about an invalid kind coming in so the `IsValid` is no longer needed.

If you decide to change it for the kind, also do so for the status.

**Review by tjerman — COMMENTED** [review]
Before you merge this, consider cleaning up your commits a bit.
I sometimes `git reset --soft` all of my commits and re-do them to make them cleaner.

I also spotted you did some chart translation commits which probably shouldn't be in this PR; can you move those out?

File: pkg/wfexec/session.go (hunk 7 above)
Comment (vicpatel) [reply pkg/wfexec/session.go]: IDE warnings; this change doesn't relate to this PR.

File: system/rest.yaml (hunk 10 above)
Comment (vicpatel) [reply system/rest.yaml]: These routes were removed and no longer needed as we have new routes for request comments.

**Comment by vicpatel** [comment]
> I also spotted you did some chart translation commits which probably shouldn't be in this PR; can you move those out?

Yeah, it was a bit of a blocker, but I kept a separate commit for it in this branch and squashed other data privacy requests related commits into one commit.
